ui/viewmodel/NetworkViewModel.kt

```diff
@@ -0,0 +1,275 @@
+            LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, TimeUnit.SECONDS.toMillis(5))
+                .setMinUpdateIntervalMillis(TimeUnit.SECONDS.toMillis(5))
+                .build()
+        } else null
+    }
+
+    fun addLocationResult(result: LocationResult) {
+        viewModelScope.launch {
+            val millis = measureTimeMillis {
+                result.locations.forEach { loc ->
+                    Log.i("Location Result", "Received $loc")
+                    val network = repository.getNetworkData()
+                    val data = GeoSignalData(
+                        sid = sessionId,
+                        timestamp = loc.time,
+                        lat = loc.latitude,
+                        lon = loc.longitude,
+                        alt = loc.altitude,
+                        networkType = network.networkType,
+                        rssi = network.rssi,
+                        sinr = network.sinr,
+                        rsrp = network.rsrp,
+                        rsrq = network.rsrq,
+                        asu = network.asu,
+                        mno = network.mno,
+                        band = network.band,
+                        mcc = network.mcc,
+                        mnc = network.mnc,
+                        tac = network.tac,
+                        cid = network.cid,
+                        sent = false
+                    )
+                    repository.insertData(data)
+                }
+            }
+            Log.i("BASELINE", "Location update + DB insert took $millis ms")
+            if (_autoSendEnabled.value) {
+                autoSend()
+            }
+        }
+    }
+
+    private fun markAsSent(data: GeoSignalData) {
+        viewModelScope.launch {
+            repository.saveData(data.copy(sent = true))
+        }
+    }
+    // overloaded mark as sent with list
+    private fun markAsSent(dataList: List<GeoSignalData>) {
+        viewModelScope.launch {
+            val updatedList = dataList.map { it.copy(sent = true) }
+            repository.saveDataList(updatedList)
+        }
+    }
+
+    fun clearDatabase() {
+        viewModelScope.launch {
+            repository.clearAllData()
+        }
+    }
+
+    fun sendUnsentData() {
+        viewModelScope.launch {
+            // Fetch the latest unsent data from the Database
+            val unsent = repository.getUnsentSignalData().firstOrNull() ?: emptyList()
+
+            val application = getApplication<Application>()
+
+            if (unsent.isEmpty()) {
+                Toast.makeText(application, "No unsent data to send", Toast.LENGTH_SHORT).show()
+                return@launch
+            }
+
+            val postService = PostService.create()
+            var allSuccess = true
+
+            for (data in unsent) {
+                val request = data.toPostRequest()
+                try {
+                    val response = postService.postGeoSignal(request)
```

Comment:
```suggestion
                    val response = postService.postGeoSignal(request) ?: return
```
Voeg een null check toe zodat je hieronder minder stappen hoeft te zetten om op null te checken.